pkg/impersonation/impersonation.go

```diff
@@ -30,46 +31,55 @@ func New(user user.Info, clusterContext *config.UserContext) Impersonator {
 	return Impersonator{user: user, clusterContext: clusterContext}
 }
 
-func (i *Impersonator) SetUpImpersonation() error {
+func (i *Impersonator) SetUpImpersonation() (*corev1.ServiceAccount, error) {
 	rules := i.rulesForUser()
+	logrus.Tracef("impersonation: checking role for user %s", i.user.GetName())
 	role, err := i.checkAndUpdateRole(rules)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	if role != nil {
-		return nil
+		return i.getServiceAccount()
 	}
+	logrus.Tracef("impersonation: creating impersonation namespace")
 	err = i.createNamespace()
 	if err != nil {
-		return err
+		return nil, err
 	}
+	logrus.Tracef("impersonation: creating role for user %s", i.user.GetName())
 	role, err = i.createRole(rules)
 	if err != nil {
-		return err
+		return nil, err
 	}
+	logrus.Tracef("impersonation: creating service account for user %s", i.user.GetName())
 	sa, err := i.createServiceAccount(role)
 	if err != nil {
-		return err
+		return nil, err
 	}
+	logrus.Tracef("impersonation: creating role binding for user %s", i.user.GetName())
 	err = i.createRoleBinding(role, sa)
 	if err != nil {
-		return err
+		return nil, err
 	}
+	logrus.Tracef("impersonation: waiting for service account to become active for user %s", i.user.GetName())
 	return i.waitForServiceAccount(sa)
 }
 
-func (i *Impersonator) GetToken() (string, error) {
-	name := ImpersonationPrefix + i.user.GetUID()
-	sa, err := i.clusterContext.Core.ServiceAccounts("").Controller().Lister().Get(impersonationNamespace, name)
-	if err != nil {
-		return "", err
-	}
+func (i *Impersonator) GetToken(sa *corev1.ServiceAccount) (string, error) {
 	if len(sa.Secrets) == 0 {
 		return "", fmt.Errorf("service account is not ready")
 	}
 	secret := sa.Secrets[0]
 	secretObj, err := i.clusterContext.Core.Secrets("").Controller().Lister().Get(impersonationNamespace, secret.Name)
 	if err != nil {
+		logrus.Tracef("impersonation: error getting service account token %s: %v", secret.Name, err)
```

Comment:
This can be wrapped in a `if logrus.GetLevel() >= logrus.TraceLevel {` so we don't do the list unnecessarily, same for the below sets of trace things.

Reply:
done